File: ruby/stack.rb
# linear structure represented by a physical stack, implements
#  push(insert item into a stack), pop (delete an item from the stack), peek

class Stack_a
  def initialize(size=0)
    @stack_a = Array.new(size)
    @top = 0
  end

  def empty?
    return @stack_a[0] == nil
  end



  def push(item)
    if (self.empty?)
       @stack_a[@top] = item
    else
      @top >= @stack_a.length ? @stack_a.push(item): @stack_a[@top + 1] = item
      @top += 1
    end

    return nil
  end

  def pop
    return "ERROR" if self.empty?

    element_to_pop = @stack_a[@top]
    @stack_a[@top] = nil
    @top -= 1 unless @top == 0

    return element_to_pop
  end

  def peek
    return @stack_a[@top]
  end

  def top
    return "Top's Index: #{@top}"
  end

  def print_stack
    @stack_a.each do |item|
      unless item == nil
        print "#{item} - "
      end
    end
    print "\n"
  end

end


# class test drivers -------------------------------------------


# nu_stack = Stack_a.new(5)
# nu_stack.push("sup")
# nu_stack.push("holmes")
# nu_stack.push("holmes")
# nu_stack.push("holmes")
# nu_stack.push("sup")
# nu_stack.push("holmes")
# puts "pip #{nu_stack.peek()}"
# nu_stack.pop()
# puts "pip #{nu_stack.peek()}"

File: python/stack.py
from array import *

class Stack:

    def __init__(self,size):
        self.stack_a = []
        if size > 0:
            self.stack_a.extend([None]*size)
        self.top = 0

    def empty(self):
        return self.stack_a[0] is None

    def printList(self):
        if self.empty():
            print "the list is empty"
        else:
            print "The list has:",
            for item in das_stack.stack_a:
                print item,
            print ""


    def push(self, item):
        if self.empty():
            self.stack_a[self.top] = item
        else:
            if self.top >= len(self.stack_a):
                self.stack_a.append(item)
            else:
                self.stack_a[self.top + 1] = item
            self.top = self.top + 1

        print item,
        print " was added to the stack"

    def pop(self):
        if self.empty():
            return "Error"

        element_to_pop = self.stack_a[self.top]
        self.stack_a[self.top] = None
        self.top -= 1

        print element_to_pop,
        print " was popped"
        return element_to_pop

    def peek(self):
        print "Peepin',",
        print self.stack_a[self.top]



das_stack = Stack(5)

das_stack.printList()

das_stack.push("lol")
das_stack.push("rofl")

das_stack.printList()

das_stack.pop()

das_stack.printList()

das_stack.peek()

# print das_stack.top
# print das_stack.stack_a[0]
# print das_stack.empty()
# print das_stack.stack_a
# print das_stack.top
